### PR TITLE
calculate price and premium inside buyCover

### PR DESCRIPTION
**Pull request by danoctavian**

### CODE REVIEW

File: contracts/modules/v2/StakingPool.sol [hunk 1]
```diff
@@ -103,6 +103,9 @@ contract StakingPool is ERC20 {
   uint public constant PARAM_PRECISION = 10_000;
   uint public constant BUCKET_SIZE = 7 days;
 
+  uint public constant PRICE_CURVE_EXPONENT = 7;
+  uint public constant MAX_PRICE_PERCENTAGE = 1e20;
```
Comment (horodnicdragos) [inline contracts/modules/v2/StakingPool.sol:107]: Rename pricePercentage to priceRatio and MAX_PRICE_PERCENTAGE to PRICE_DENOMINATOR

File: contracts/modules/v2/StakingPool.sol [hunk 2]
```diff
@@ -354,4 +367,53 @@ contract StakingPool is ERC20 {
...
+  function calculatePriceIntegralAtPoint(
+    uint basePrice,
+    uint activeCover,
+    uint capacity
+  ) public pure returns (uint) {
+    uint actualPrice = basePrice * activeCover;
+    for (uint i = 0; i < PRICE_CURVE_EXPONENT; i++) {
+      actualPrice = actualPrice * activeCover / capacity;
+    }
+    actualPrice = actualPrice / 8 + basePrice * activeCover;
```
Comment (horodnicdragos) [inline contracts/modules/v2/StakingPool.sol:414]: Where does the division by 8 come from?

File: contracts/modules/v2/StakingPool.sol [hunk 3]
```diff
@@ -354,4 +367,53 @@ contract StakingPool is ERC20 {
...
+  }
+
+  function calculatePrice(
+    uint amount,
+    uint basePrice,
+    uint activeCover,
+    uint capacity
+  ) public pure returns (uint) {
+
+    return (calculatePriceIntegralAtPoint(
```
Comment (horodnicdragos) [inline contracts/modules/v2/StakingPool.sol:393]: I think I am out of context here. I'm pretty sure there's a doc with the formulas that might be worth adding here as comment to help people that read the code understand what is going on.

File: contracts/modules/v2/StakingPool.sol [hunk 4]
```diff
@@ -276,6 +279,16 @@ contract StakingPool is ERC20 {
       product.lastBucket = uint16(lastBucket);
       product.activeCoverAmount = uint96(activeCoverAmount + coverAmount);
     }
+
+    // price calculation
+    uint pricePercentage = calculatePrice(
+      coverAmount,
+      basePrice,
```
Comment (shark0der) [inline contracts/modules/v2/StakingPool.sol:286]: What's `basePrice`?

File: contracts/modules/cover/Cover.sol [hunk 5]
```diff
@@ -383,36 +383,14 @@ contract Cover is ICover, MasterAwareV2 {
         /*uint8 decimals*/,
         /*bool deprecated*/
       ) = pool().assets(payoutAssetIndex);
-      
+
```
Comment (horodnicdragos) [inline contracts/modules/cover/Cover.sol:386]: This is a comment

**Pull request merged**